core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/VectorCodec.java

```diff
@@ -156,4 +116,195 @@ public CqlVector<SubtypeT> parse(@Nullable String value) {
+      Iterator<SubtypeT> values = value.iterator();
+      int allValueBuffsSize = 0;
+      for (int i = 0; i < cqlType.getDimensions(); ++i) {
+        ByteBuffer valueBuff;
+        SubtypeT valueObj;
+
+        try {
+          valueObj = values.next();
+        } catch (NoSuchElementException nsee) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "Not enough elements; must provide elements for %d dimensions",
+                  cqlType.getDimensions()));
+        }
+
+        try {
+          valueBuff = this.subtypeCodec.encode(valueObj, protocolVersion);
+        } catch (ClassCastException e) {
+          throw new IllegalArgumentException("Invalid type for element: " + valueObj.getClass());
+        }
+        if (valueBuff == null) {
+          throw new NullPointerException("Vector elements cannot encode to CQL NULL");
+        }
+        int elementSize = valueBuff.limit();
+        allValueBuffsSize += elementSize + VIntCoding.computeVIntSize(elementSize);
+        valueBuff.rewind();
+        valueBuffs[i] = valueBuff;
+      }
+
+      // if too many elements, throw
+      if (values.hasNext()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Too many elements; must provide elements for %d dimensions",
+                cqlType.getDimensions()));
+      }
+
+      /* Since we already did an early return for <= 0 dimensions above */
+      assert valueBuffs.length > 0;
+      ByteBuffer rv = ByteBuffer.allocate(allValueBuffsSize);
+      for (int i = 0; i < cqlType.getDimensions(); ++i) {
+        VIntCoding.writeUnsignedVInt32(valueBuffs[i].remaining(), rv);
+        rv.put(valueBuffs[i]);
+      }
+      rv.flip();
+      return rv;
+    }
+
+    @Override
+    public CqlVector<SubtypeT> decode(
+        @Nullable ByteBuffer bytes, @NonNull ProtocolVersion protocolVersion) {
+      if (bytes == null || bytes.remaining() == 0) {
+        return null;
+      }
+
+      ByteBuffer input = bytes.duplicate();
+      List<SubtypeT> rv = new ArrayList<SubtypeT>(cqlType.getDimensions());
+      for (int i = 0; i < cqlType.getDimensions(); ++i) {
+        int size = VIntCoding.getUnsignedVInt32(input, input.position());
+        input.position(input.position() + VIntCoding.computeUnsignedVIntSize(size));
+
+        ByteBuffer value;
+        if (size < 0) {
+          value = null;
+        } else {
+          value = input.duplicate();
+          value.limit(value.position() + size);
+          input.position(input.position() + size);
+        }
+        rv.add(subtypeCodec.decode(value, protocolVersion));
+      }
+      // if too many elements, throw
+      if (input.hasRemaining()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Too many elements; must provide elements for %d dimensions",
+                cqlType.getDimensions()));
+      }
+
+      return CqlVector.newInstance(rv);
+    }
```

Comment:
I feel like there's a lot of code duplication between the 2 "proxy" codecs, I'd suggest either making it a single proxy codec (or just implementing it in the vector codec itself) or creating some methods (possibly an abstract class?) that can be reused by both proxy codecs. The only difference between both of them is that on one codec you have to read/write the length before the actual data to get the element size and on the other one you already have that element size so you can just skip the read/write of that size.